src/data/upload.ts

```diff
@@ -143,25 +149,27 @@ function stringify(payload: EncodedPayload): string {
     return `{"e":${payload.e},"d":${payload.d}}`;
 }
 
-function send(data: string, sequence: number = null, last: boolean = false): void {
+function send(data: string, sequence: number, last: boolean): void {
     // Upload data if a valid URL is defined in the config
     if (config.url.length > 0) {
-        if (last && "sendBeacon" in navigator) {
+        // Upload data using sendBeacon only when it's the last payload and the data size is within safe limits.
+        // Most browsers in the wild will reject the sendBeacon call if data being transferred is over 64KB (BEACON_BYTES_LIMIT).
+        if (last && data.length < BEACON_BYTES_LIMIT && "sendBeacon" in navigator) {
             navigator.sendBeacon(config.url, data);
```

Comment:
The second beacon call returns a Boolean if it succeeds. Could you change this to use that value and then fall back to the XHR if it return false? That seems safer than a hardcoded assumption of data size. See: return value in: [https://www.w3.org/TR/beacon/#sec-sendBeacon-method](https://www.w3.org/TR/beacon/#sec-sendBeacon-method).

Reply:
We invoke beacon call when we are almost ready to leave the page - on unload. At that point we can't wait and then reissue the request because browser would've already unloaded us from the memory by then. We have one chance to fire the request and if we don't get it right, we lost the chance. I could go with sync xhr but that can have performance implications so not going there and keeping it simple.